Tekton Extension fails to deserialize parameter values

Hi,

first of all, thanks for the tekton extension! Since Tekton 0.6.0, param values now also support Arrays in addition to Strings (https://github.com/tektoncd/pipeline/pull/1080). This seems to break the usage of `tknClient().pipelines().load(is).createOrReplace()` with the following input (simplified)

```
apiVersion: tekton.dev/v1alpha1
kind: Pipeline
metadata:
  name: pipeline
spec:
  tasks:
    - name: task-with-params
      params:
        - name: name
          value: param-value
```
The load method throws
```
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `io.fabric8.tekton.pipeline.v1alpha1.ArrayOrString` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('value')
 at [Source: (BufferedInputStream); line: 10, column: 18] (through reference chain: io.fabric8.tekton.pipeline.v1alpha1.Pipeline["spec"]->io.fabric8.tekton.pipeline.v1alpha1.PipelineSpec["tasks"]->java.util.ArrayList[0]->io.fabric8.tekton.pipeline.v1alpha1.PipelineTask["params"]->java.util.ArrayList[0]->io.fabric8.tekton.pipeline.v1alpha1.Param["value"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1343)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1032)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:371)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromString(StdValueInstantiator.java:323)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromString(BeanDeserializerBase.java:1373)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:171)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:161)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4014)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3078)
	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:241)
	... 35 more
```

As I understand, the tekton extension is generated from the CRD definitions (?!). Can this be adjusted to support deserializing `io.fabric8.tekton.pipeline.v1alpha1.ArrayOrString`?

I´m using the latest version from master branch for this.

Thanks, Fabian

--- a/extensions/tekton/model/pom.xml
+++ b/extensions/tekton/model/pom.xml
@@ -94,22 +94,6 @@
     <build>
       <plugins>
         <plugin>
-          <groupId>com.mysema.maven</groupId>
-          <artifactId>apt-maven-plugin</artifactId>
-          <version>1.1.3</version>
-          <executions>
-            <execution>
-              <goals>
-                <goal>process</goal>
-              </goals>
-              <configuration>
-                <outputDirectory>target/generated-sources</outputDirectory>
-                <processor>io.sundr.builder.internal.processor.BuildableProcessor</processor>
-              </configuration>
-            </execution>
-          </executions>
-        </plugin>
-        <plugin>
           <groupId>org.jsonschema2pojo</groupId>
           <artifactId>jsonschema2pojo-maven-plugin</artifactId>
           <version>${jsonschema2pojo.version}</version>
@@ -139,6 +123,29 @@
               <version>${project.version}</version>
             </dependency>
           </dependencies>
+        </plugin>
+        <plugin>
+          <artifactId>maven-antrun-plugin</artifactId>
+          <version>1.8</version>
+          <executions>
+            <execution>
+              <phase>generate-sources</phase>
+              <configuration>
+                <target>
+                  <echo>removing the duplicate generated calss</echo>
+                  <delete file="${basedir}/target/generated-sources/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrString.java" verbose="true" />
+                  <delete verbose="true">
+                    <fileset dir="${basedir}/target/generated-sources">
+                      <include name="*.java" />
+                    </fileset>
+                  </delete>
+                </target>
+              </configuration>
+              <goals>
+                <goal>run</goal>
+              </goals>
+            </execution>
+          </executions>
         </plugin>
         <plugin>
           <groupId>org.codehaus.mojo</groupId>

--- a/extensions/tekton/model/src/main/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrString.java
+++ b/extensions/tekton/model/src/main/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrString.java
@@ -1,3 +1,18 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package io.fabric8.tekton.pipeline.v1alpha1;
 
 

--- a/extensions/tekton/model/src/main/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrString.java
+++ b/extensions/tekton/model/src/main/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrString.java
@@ -1,0 +1,170 @@
+package io.fabric8.tekton.pipeline.v1alpha1;
+
+
+import com.fasterxml.jackson.annotation.*;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.ObjectCodec;
+import com.fasterxml.jackson.databind.*;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import io.fabric8.kubernetes.api.model.Doneable;
+import io.fabric8.kubernetes.api.model.KubernetesResource;
+import io.sundr.builder.annotations.Buildable;
+import io.sundr.builder.annotations.Inline;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+import java.io.IOException;
+import java.util.*;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+  "arrayVal",
+  "stringVal",
+  "type"
+})
+@JsonDeserialize(using = ArrayOrString.Deserializer.class)
+@JsonSerialize(using = ArrayOrString.Serializer.class)
+@ToString
+@EqualsAndHashCode
+@Buildable(editableEnabled = false, validationEnabled = false, generateBuilderPackage = false, builderPackage = "io.fabric8.kubernetes.api.builder", inline = @Inline(type = Doneable.class, prefix = "Doneable", value = "done"))
+public class ArrayOrString implements KubernetesResource {
+
+  private static final String TYPE_STRING = "string";
+  private static final String TYPE_ARRAY = "array";
+
+  @JsonProperty("arrayVal")
+  private List<String> arrayVal = new ArrayList<String>();
+
+  @JsonProperty("stringVal")
+  private String stringVal;
+
+  @JsonProperty("type")
+  private String type;
+
+  @JsonIgnore
+  private Map<String, Object> additionalProperties = new HashMap<String, Object>();
+
+  /**
+   * No args constructor for use in serialization
+   */
+  public ArrayOrString() {
+  }
+
+  //Builders are generated for the first non-empty constructor found.
+  public ArrayOrString(List<String> arrayVal, String stringVal, String type, Map<String, Object> additionalProperties) {
+    this.arrayVal = arrayVal;
+    this.stringVal = stringVal;
+    this.type = type;
+    this.additionalProperties = additionalProperties;
+  }
+
+  public ArrayOrString(String stringVal) {
+    this(null, stringVal, TYPE_STRING, new HashMap<>());
+  }
+
+  public ArrayOrString(List<String> arrayVal) {
+    this(arrayVal, null, TYPE_ARRAY, new HashMap<>());
+  }
+
+  @JsonProperty("arrayVal")
+  public List<String> getArrayVal() {
+    return arrayVal;
+  }
+
+  @JsonProperty("arrayVal")
+  public void setArrayVal(List<String> arrayVal) {
+    this.arrayVal = arrayVal;
+  }
+
+  @JsonProperty("stringVal")
+  public String getStringVal() {
+    return stringVal;
+  }
+
+  @JsonProperty("stringVal")
+  public void setStringVal(String stringVal) {
+    this.stringVal = stringVal;
+  }
+
+  @JsonProperty("type")
+  public String getType() {
+    return type;
+  }
+
+  @JsonProperty("type")
+  public void setType(String type) {
+    this.type = type;
+  }
+
+  @JsonAnyGetter
+  public Map<String, Object> getAdditionalProperties() {
+    return this.additionalProperties;
+  }
+
+  @JsonAnySetter
+  public void setAdditionalProperty(String name, Object value) {
+    this.additionalProperties.put(name, value);
+  }
+
+  public static class Serializer extends JsonSerializer<ArrayOrString> {
+
+    @Override
+    public void serialize(ArrayOrString value, JsonGenerator jgen, SerializerProvider provider) throws IOException, JsonProcessingException {
+      if (value != null) {
+        if (value.getType() == null) {
+          String stringVal = value.getStringVal();
+          if (stringVal != null) {
+            jgen.writeString(stringVal);
+          } else {
+            List<String> arrayVal = value.getArrayVal();
+            if (arrayVal != null) {
+              writeArray(value, jgen);
+            } else {
+              jgen.writeNull();
+            }
+          }
+        } else if (Objects.equals(value.getType(), TYPE_STRING)) {
+          jgen.writeString(value.stringVal);
+        } else if (Objects.equals(value.getType(), TYPE_ARRAY)) {
+          writeArray(value, jgen);
+        } else {
+          jgen.writeNull();
+        }
+      } else {
+        jgen.writeNull();
+      }
+    }
+
+    private void writeArray(ArrayOrString value, JsonGenerator jgen) throws IOException {
+      jgen.writeStartArray(value.getArrayVal().size());
+      for (String n : value.getArrayVal()) {
+        jgen.writeString(n);
+      }
+      jgen.writeEndArray();
+    }
+
+  }
+
+  public static class Deserializer extends JsonDeserializer<ArrayOrString> {
+
+    @Override
+    public ArrayOrString deserialize(JsonParser jsonParser, DeserializationContext ctxt) throws IOException {
+      ObjectCodec oc = jsonParser.getCodec();
+      JsonNode node = oc.readTree(jsonParser);
+      ArrayOrString arrayOrString;
+      if (node.isArray()) {
+        List<String> elements = new ArrayList<>();
+        node.elements().forEachRemaining(n -> elements.add(n.asText()));
+        arrayOrString = new ArrayOrString(elements);
+      } else {
+        arrayOrString = new ArrayOrString(node.asText());
+      }
+      return arrayOrString;
+    }
+
+  }
+
+}

--- a/extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrStringTest.java
+++ b/extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrStringTest.java
@@ -1,3 +1,18 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package io.fabric8.tekton.pipeline.v1alpha1;
 
 import com.fasterxml.jackson.databind.ObjectMapper;

--- a/extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrStringTest.java
+++ b/extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1alpha1/ArrayOrStringTest.java
@@ -1,0 +1,27 @@
+package io.fabric8.tekton.pipeline.v1alpha1;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class ArrayOrStringTest {
+
+  @Test
+  public void testArrayOrStringJson() throws IOException {
+    ArrayOrString is = new ArrayOrString("string-value");
+    ObjectMapper mapper = new ObjectMapper();
+    String json = mapper.writeValueAsString(is);
+    ArrayOrString is2 = mapper.readValue(json, ArrayOrString.class);
+    assertEquals(is, is2);
+
+    is = new ArrayOrString(Arrays.asList("a", "b"));
+    json = mapper.writeValueAsString(is);
+    is2 = mapper.readValue(json, ArrayOrString.class);
+    assertEquals(is, is2);
+  }
+
+}

--- a/extensions/tekton/tests/src/test/java/io/fabric8/tekton/test/crud/PipelineCrudTest.java
+++ b/extensions/tekton/tests/src/test/java/io/fabric8/tekton/test/crud/PipelineCrudTest.java
@@ -18,18 +18,22 @@ package io.fabric8.tekton.test.crud;
 import io.fabric8.tekton.pipeline.v1alpha1.*;
 import io.fabric8.tekton.client.TektonClient;
 import io.fabric8.tekton.mock.TektonServer;
-
+import io.fabric8.tekton.pipeline.v1alpha1.Param;
+import io.fabric8.tekton.pipeline.v1alpha1.Pipeline;
+import io.fabric8.tekton.pipeline.v1alpha1.PipelineList;
 import org.junit.Rule;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import java.io.ByteArrayInputStream;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
 
 @EnableRuleMigrationSupport
 public class PipelineCrudTest {
-  
+
   @Rule
   public TektonServer server = new TektonServer(true, true);
 
@@ -64,4 +68,30 @@ public class PipelineCrudTest {
     Boolean deleted = client.pipelines().inNamespace("ns3").withName("pipeline3").delete();
     assertTrue(deleted);
   }
+
+  @Test
+  public void shouldLoadAPipelineWithParams() {
+    TektonClient client = server.getTektonClient();
+
+    String pipelineDefinition = String.join("\n", Arrays.asList(
+      "apiVersion: tekton.dev/v1alpha1",
+      "kind: Pipeline",
+      "metadata:",
+      "  name: pipeline4",
+      "spec:",
+      "  tasks:",
+      "    - name: task-with-params",
+      "      params:",
+      "        - name: name",
+      "          value: param-value"
+    ));
+
+    Pipeline p = client.pipelines().inNamespace("ns4").load(new ByteArrayInputStream(pipelineDefinition.getBytes())).createOrReplace();
+
+    final List<Param> taskParams = p.getSpec().getTasks().get(0).getParams();
+    assertEquals(1, taskParams.size());
+    assertEquals("name", taskParams.get(0).getName());
+    assertEquals("param-value", taskParams.get(0).getValue().getStringVal());
+  }
+
 }

